File: templates/js/junit-to-csv.mjs
#!/usr/bin/env node
// Converts a Playwright JUnit report into the framework's results CSV:
//   id,type,role,route,expected,actual,verdict,ms
//
// Standard library only (node:fs, node:path) — no xml2js, no fast-xml-parser.
// The JUnit shape here is narrow and well known (Playwright's own writer), so
// a small regex walk is more honest than pulling in a parser for it.
//
// Usage: node junit-to-csv.mjs tests/results/junit.xml tests/results/run-<ts>.csv
import { readFileSync, writeFileSync } from "node:fs";

const [, , inFile, outFile] = process.argv;
if (!inFile || !outFile) {
  console.error("usage: junit-to-csv.mjs <junit.xml> <out.csv>");
  process.exit(1);
}

const xml = readFileSync(inFile, "utf8");

// The case id is the leading token of the test name/title, e.g.
// "INV-014 create an invoice" -> "INV-014". This is the js.md convention:
// hyphenated ids survive as-is inside a JS string, so no re-encoding is
// needed the way Python/Java/.NET identifiers require.
// Multi-segment prefixes are real: `tf.sh rbac` emits RBAC-USER-002.
const ID_RE = /([A-Z][A-Z0-9]*(?:-[A-Z][A-Z0-9]*)*-\d+)/;

function attr(tag, name) {
  const m = tag.match(new RegExp(name + '="([^"]*)"'));
  return m ? m[1] : "";
}

function csvField(v) {
  const s = String(v ?? "");
  return /[",\n]/.test(s) ? '"' + s.replace(/"/g, '""') + '"' : s;
}

const rows = ["id,type,role,route,expected,actual,verdict,ms"];

// Each <testcase ...> is either self-closing (pass) or wraps a <failure>/
// <error>/<skipped> child. Walk them as whole elements, not line by line —
// a JUnit file is not guaranteed to be one testcase per line.
// The quantifier MUST be lazy. Greedy `[^>]*` consumes the `/` of a
// self-closing tag, so the `\/>` branch can no longer match and the pattern
// falls through to `>...</testcase>` -- swallowing the NEXT case and pairing
// this case with that one's failure.
const caseRe = /<testcase\b[^>]*?(?:\/>|>[\s\S]*?<\/testcase>)/g;
for (const tag of xml.match(caseRe) ?? []) {
  const name = attr(tag, "name");
  const idMatch = name.match(ID_RE);
  if (!idMatch) continue; // not one of ours (e.g. a beforeAll hook)
  const id = idMatch[1];

  const ms = Math.round(parseFloat(attr(tag, "time") || "0") * 1000);
  let verdict = "PASS";
  if (/<error\b/.test(tag)) verdict = "ERROR";
  else if (/<failure\b/.test(tag)) verdict = "FAIL";
  else if (/<skipped\b/.test(tag)) verdict = "SKIP";

  const msgMatch = tag.match(/<(?:failure|error)\b[^>]*message="([^"]*)"/);
  const actual = msgMatch ? msgMatch[1] : verdict === "PASS" ? "ok" : "";

  // type/role/route are not in JUnit output at all — the spec-writer agent
  // stamped only the id into the test name. Leave them for the caller to
  // fold back in from testcases.csv by id if it needs them; this adapter's
  // job is verdict + timing, which is all `tf.sh setmany` consumes.
  rows.push([id, "ui", "", "", "", csvField(actual), verdict, ms].join(","));
}

writeFileSync(outFile, rows.join("\n") + "\n");
console.error(`junit-to-csv: wrote ${rows.length - 1} rows to ${outFile}`);

File: templates/python/junit_to_csv.py
#!/usr/bin/env python3
"""Convert pytest's JUnit report into the framework's results CSV:
id,type,role,route,expected,actual,verdict,ms

Standard library only (xml.etree.ElementTree, csv, re, sys) -- no lxml, no
extra dependency of any kind.

Usage: python junit_to_csv.py tests/results/junit.xml tests/results/run-<ts>.csv
"""
import csv
import re
import sys
import xml.etree.ElementTree as ET

# skills/codegen/references/python.md's naming convention: a case id
# like INV-014 is encoded in the test function name as INV__014 (hyphen ->
# double underscore, since a hyphen is not a legal Python identifier). This
# reverses it.
# Multi-segment prefixes are real: `tf.sh rbac` emits RBAC-USER-002,
# which codegen writes as test_RBAC_USER__002_...
ID_RE = re.compile(r"test_([A-Z][A-Z0-9]*(?:_[A-Z][A-Z0-9]*)*)__(\d+)")


def recover_id(name: str) -> str | None:
    m = ID_RE.search(name)
    if not m:
        return None
    # Single underscores separate prefix segments and must become hyphens
    # again: test_RBAC_USER__002 -> RBAC-USER-002.
    return f"{m.group(1).replace('_', '-')}-{m.group(2)}"


def main(in_path: str, out_path: str) -> None:
    tree = ET.parse(in_path)
    root = tree.getroot()
    # pytest's junit-xml wraps one or more <testsuite>; be tolerant of both
    # a bare <testsuite> root and a <testsuites> wrapper.
    suites = [root] if root.tag == "testsuite" else list(root)

    rows = []
    for suite in suites:
        for case in suite.findall("testcase"):
            name = case.get("name", "")
            case_id = recover_id(name)
            if case_id is None:
                continue  # not one of ours (e.g. a fixture-only node)

            ms = round(float(case.get("time", "0")) * 1000)
            failure = case.find("failure")
            error = case.find("error")
            skipped = case.find("skipped")
            if error is not None:
                verdict, actual = "ERROR", error.get("message", "")
            elif failure is not None:
                verdict, actual = "FAIL", failure.get("message", "")
            elif skipped is not None:
                verdict, actual = "SKIP", skipped.get("message", "")
            else:
                verdict, actual = "PASS", "ok"

            # type/role/route are not present in JUnit output -- only the id
            # was stamped into the test name. The caller folds those back in
            # from testcases.csv by id; this adapter's job is verdict+timing.
            rows.append([case_id, "ui", "", "", "", actual, verdict, ms])

    with open(out_path, "w", newline="") as f:
        w = csv.writer(f)
        w.writerow(["id", "type", "role", "route", "expected", "actual", "verdict", "ms"])
        w.writerows(rows)

    print(f"junit_to_csv: wrote {len(rows)} rows to {out_path}", file=sys.stderr)


if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("usage: junit_to_csv.py <junit.xml> <out.csv>", file=sys.stderr)
        sys.exit(1)
    main(sys.argv[1], sys.argv[2])
